server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,275 @@
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING;
+import static org.elasticsearch.health.node.selection.HealthNodeTaskExecutor.ENABLED_SETTING;
+
+/**
+ * Keeps the health metadata in the cluster state up to date. It listens to master elections and changes in the disk thresholds.
+ */
+public class HealthMetadataService {
+
+    private static final Logger logger = LogManager.getLogger(HealthMetadataService.class);
+
+    private final ClusterService clusterService;
+    private final ClusterStateListener clusterStateListener;
+    private final DiskHealthMetadataMonitor diskHealthMetadataMonitor;
+    private volatile boolean enabled;
+    private volatile boolean publishedAfterElection = false;
+
+    private final ClusterStateTaskExecutor<UpdateHealthMetadataTask> taskExecutor = new UpdateHealthMetadataTask.Executor();
+
+    public HealthMetadataService(ClusterService clusterService, Settings settings) {
+        this.clusterService = clusterService;
+        this.diskHealthMetadataMonitor = new DiskHealthMetadataMonitor(
+            settings,
+            clusterService.getClusterSettings(),
+            this::updateHealthMetadata
+        );
+        this.clusterStateListener = this::updateHealthMetadataIfNecessary;
+        this.enabled = ENABLED_SETTING.get(settings);
+        if (this.enabled) {
+            this.clusterService.addListener(clusterStateListener);
+        }
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ENABLED_SETTING, this::enable);
+    }
+
+    private void enable(boolean enabled) {
+        this.enabled = enabled;
+        if (this.enabled) {
+            clusterService.addListener(clusterStateListener);
+            updateHealthMetadata();
+        } else {
+            clusterService.removeListener(clusterStateListener);
+            publishedAfterElection = false;
+        }
+    }
+
+    private void updateHealthMetadataIfNecessary(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            if (event.localNodeMaster() && publishedAfterElection == false) {
+                submitHealthMetadata("health-metadata-update-master-election");
+            }
+            // If the node is not the elected master anymore
```

Comment:
Yes, functionally, that's exactly what I was going for. I think I cannot reuse the same code because we wait to check for the version of the cluster `event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)`. If I am not mistaken this happens after the master node is elected so we do not publish the health metadata during the start up phase of a cluster. 

Do you have any suggestions to capture this scenario?

Reply:
I will try to use `LocalNodeMasterListener`. Maybe it will produce some cleaner code.

Reply:
No, that wouldn't work because still need to check the version `event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)`.

Reply:
I am working on something, I will ping you back when it's ready :)